application/apps/indexer/session/src/handlers/search.rs

```diff
@@ -5,46 +5,95 @@ use crate::{
 };
 use indexer_base::progress::Severity;
 use log::debug;
-use processor::search::{FilterStats, SearchFilter, SearchHolder};
-use std::path::PathBuf;
+use processor::search::{FilterStats, SearchFilter, SearchHolder, SearchResults};
+use tokio::{
+    sync::mpsc::{channel, Receiver, Sender},
+    task,
+    time::{timeout, Duration},
+};
+
+const TRACKING_INTERVAL_MS: u64 = 250;
 
 pub async fn handle(
     operation_api: &OperationAPI,
-    target_file: PathBuf,
     filters: Vec<SearchFilter>,
     state: SessionStateAPI,
 ) -> OperationResult<SearchOperationResult> {
     debug!("RUST: Search operation is requested");
+    state.drop_search().await?;
     if filters.is_empty() {
         debug!("RUST: Search will be dropped. Filters are empty");
-        state.drop_search().await?;
         operation_api.emit(CallbackEvent::SearchUpdated(0));
         Ok(Some(SearchOperationResult {
             found: 0,
             stats: FilterStats::new(vec![]),
         }))
     } else {
-        let search_holder = SearchHolder::new(&target_file, filters.iter());
-        let search_results = match search_holder.execute_search() {
-            Ok((file_path, matches, stats)) => Ok((file_path, matches.len(), matches, stats)),
-            Err(err) => {
-                return Err(NativeError {
-                    severity: Severity::ERROR,
-                    kind: NativeErrorKind::OperationSearch,
-                    message: Some(format!("Fail to execute search. Error: {}", err)),
-                });
-            }
+        let mut search_holder = state.get_search_holder().await?;
+        search_holder.set_filters(&mut filters.iter());
+        let search_res_file = search_holder.out_file_path.clone();
+        let (tx_result, mut rx_result): (
+            Sender<(SearchHolder, SearchResults)>,
+            Receiver<(SearchHolder, SearchResults)>,
+        ) = channel(1);
+        task::spawn(async move {
+            let search_results = search_holder.execute_search();
+            if tx_result
+                .send((search_holder, search_results))
+                .await
+                .is_ok()
+            {}
+        });
+        let search_results = loop {
+            match timeout(
+                Duration::from_millis(TRACKING_INTERVAL_MS as u64),
+                rx_result.recv(),
+            )
+            .await
+            {
+                Ok(recv_results) => {
+                    break match recv_results {
```

Comment:
break match? never seen this before...what does it do?

Reply:
sorry, "copy/paste" mistake

Reply:
that's a cool feature of rust, I like it
https://doc.rust-lang.org/rust-by-example/flow_control/loop/return.html
it's a way to return value from a `loop`. It's useful to determine the reason of breaking of loop: does happen because of error or it's expectable "exit"

Reply:
yeah and `break match` here because I'm "overwriting" not only Err(...), but also Ok(...) hands.

Reply:
probably with mapping it looks more clean
```
                    break recv_results.map_or(
                        Err(NativeError {
                            severity: Severity::ERROR,
                            kind: NativeErrorKind::OperationSearch,
                            message: Some("Fail to receive search results".to_string()),
                        }),
                        |(search_holder, search_results)| {
                            search_results
                                .map(|(file_path, matches, stats)| {
                                    (file_path, matches.len(), matches, stats, search_holder)
                                })
                                .map_err(|err| NativeError {
                                    severity: Severity::ERROR,
                                    kind: NativeErrorKind::OperationSearch,
                                    message: Some(format!(
                                        "Fail to execute search. Error: {}",
                                        err
                                    )),
                                })
                        },
                    );

```